add volumeClaimTemplate to zookeeper statefulSet

jbianquetti-nami opened the pull request:
Fixes https://github.com/kubeless/kubeless/issues/320

jbianquetti-nami commented:
Tests fail due to a problem while mounting the volume:
`
pods/zoo-0: Error executing 'postInstallation': EACCES: permission denied, mkdir '/bitnami/zookeeper/conf'`

Investigating...

jbianquetti-nami commented:
I've been looking for the underlying issue about why tests fail when I added the volumeClaimTemplate definition. 

I've tried to run the definition on minikube and GKE. I've been able to run in GKE by adding this to the template:
```
securityContext:
    runAsUser: 1001
    fsGroup: 1001
```
But it doesn't work for minikube.  I tried to circumvent the problem using an initContainer to change permission on the volume, but it failed too, and it does not work on Openshift.

 That was due to minikube have an issue with permissions in directories provisioned by hostPath provisioner, which are only writeable by root (issue now closed), but the version is not yet released. See:  https://github.com/kubernetes/minikube/issues/1990

So, I think it may worth to wait until the next release of minikube to apply this PR and try to pass the tests again.

jbianquetti-nami commented:
I built minikube from the source (including the fix for  kubernetes/minikube#1990) and test successfully with this config added for the Zookeeper yaml:
```
      securityContext:
        runAsUser: 1001
        fsGroup: 1001
  volumeClaimTemplates:
  - metadata:
      name: zookeeper
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi
```
So this PR may work for minikube > v0.22.2

andresmgot commented:
@jbianquetti-nami you can update the minikube version in the `cluster-up-minikube.sh` script in order to pass the checks.

jbianquetti-nami commented:
Yes, but there's not a Minikube release including the bug fix   :)

andresmgot commented:
Ah, sorry, I thought it was already released

sebgoa commented:
and what is the issue on openshift ?

jbianquetti-nami commented:
@sebgoa the issue was related to the HostPath provisioner, so only fails on `minikube` or `minishift`.  Any other cluster provider will work.   

The PR is blocked by `test_kubeless_function_update` tests. I'm debugging why this test fails.

jbianquetti-nami commented:
Changes included:
* Updated Kafka container to bitnami/kafka@sha256:44fbc18518e7028ce87d4bec276c54b6a2cd2c2ece0c8575418d35cad0270a9c. Now the Kafka topics stores the data in the right filesystem path
* Added VolumeClaimTemplate to Zookeeper, to persist ZK information between reboots
* Added securityContext to Kafka and Zookeeper containers in `kubeless.jsonnet`, to allow USER 1001 write into volumes
* Updated Minikube to v0.22.3 to get fixes about securityContext in the HostPath provisioner
* Add `--tail=100` to `kubeless/script/libtest.bash`, to correct a grep check that makes test unstable
* Uses `bitnami/kubeless-controller` and `kubeless` cli  v0.2.3 to fix tests issues while updating functions.

jbianquetti-nami commented:
PR closed and opened again for testing if TravisCI tests are finally stable. 
Tests passed OK.  Ready to merge.

sebgoa commented:
@jbianquetti-nami can you rebase

jbianquetti-nami commented:
I'll close this without merging and open a new one to make changes more clear because task name and changes applied are not directly traceable.
